xtm/xtm_api.cc

```diff
@@ -0,0 +1,354 @@
+			close(queue->producer_write_fd);
+	}
+close_consumer_fds:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_notify_producer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->producer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_notification_expected, true,
+			 __ATOMIC_SEQ_CST);
+
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+error:
+	errno = ENOBUFS;
+	return -1;
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+unsigned
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.end();
+	return cnt;
+}
+
+int
+xtm_queue_push_ptr(struct xtm_queue *queue, void *ptr)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.ptr = ptr;
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) == 0)
+		goto error;
+
+	__atomic_store_n(&queue->is_notification_expected, true,
+			 __ATOMIC_SEQ_CST);
+
+	if (queue->queue.put(&xtm_msg, 1) != 0)
+		return 0;
+
+error:
+	errno = ENOBUFS;
+	return -1;
+}
+
+unsigned
+xtm_queue_pop_ptrs(struct xtm_queue *queue, void **ptr_array,
+		   unsigned ptr_array_size)
```

Comment:
I do not like "size" which is not in bytes. "ptr_array_items"?

Reply:
`items` would fit for the items themselves, not for the count. For count you might want to use `count`, if `size` looks wrong to you.

Reply:
Whatever. Just not "size".
"number_of_ptrs" is also not that bad.

Reply:
Fixed to ptr_array_count